controllers/auth_controller.go

```diff
@@ -0,0 +1,195 @@
+//	@Failure      500  {object}  V1_API_RESPONSE_USERS
+//	@Router       /signup [post]
+func Signup(c *gin.Context) {
+	var response types.V1_API_RESPONSE_AUTH
+	var status int
+	var ctx, cancel = context.WithTimeout(context.Background(), 100*time.Second)
+	defer cancel()
+	var uInput types.UserSignupInput
+
+	if err := c.BindJSON(&uInput); err != nil {
+		status = http.StatusBadRequest
+		response.Message = err.Error()
+		response.Status = status
+		c.JSON(status, response)
+		return
+	}
+
+	count, err := models.CountUsersByEmail(ctx, uInput.Email)
+	if err != nil {
+		log.Println("ERROR: ", err.Error())
+		status = http.StatusInternalServerError
+		response.Message = "Internal server error when checking if user exists"
+		response.Status = status
+		c.JSON(status, response)
+		return
+	}
+
+	if count > 0 {
+		log.Println("DUPLICATE USER SIGNUP ATTEMPT FOR EMAIL: ", uInput.Email)
+		status = http.StatusUnprocessableEntity
+		response.Message = "A user with this email address already exists."
+		response.Status = status
+		c.JSON(status, response)
+		return
+	}
+
+	verifyPwResult := helper.VerifyPasswordComplexity(uInput.Password, 2, 2, 2, 8)
+	if !verifyPwResult.HasExpectedDigitCt || !verifyPwResult.HasExpectedSpecialCaseCt || !verifyPwResult.HasExpectedUpperCaseCt || !verifyPwResult.HasMinLength {
+		var b strings.Builder
+		b.WriteString("Password failed complexity requirement(s): ")
+		if !verifyPwResult.HasExpectedDigitCt {
+			b.WriteString("must have 2 or more digits; ")
+		}
+		if !verifyPwResult.HasExpectedSpecialCaseCt {
+			b.WriteString("must have 2 or more special characters; ")
+		}
+		if !verifyPwResult.HasExpectedUpperCaseCt {
+			b.WriteString("must have 2 or more capital letters; ")
+		}
+		if !verifyPwResult.HasMinLength {
+			b.WriteString("must be at least 8 characters in length")
+		}
+		status = http.StatusUnprocessableEntity
+		response.Status = status
+		response.Message = b.String()
+		c.JSON(status, response)
+		return
+	}
+
+	hashedPassword := helper.HashPassword(uInput.Password)
+	var newUser = models.User{}
+	newUser.FirstName = uInput.FirstName
+	newUser.LastName = uInput.LastName
+	newUser.Email = uInput.Email
+	newUser.Password = hashedPassword
+
+	createUserInput := []models.User{newUser}
+	err = models.CreateUsers(ctx, &createUserInput)
+	if err != nil {
+		log.Println("ERROR: ", err.Error())
+		status = http.StatusInternalServerError
+		response.Message = "Internal server error while creating user"
+		response.Status = status
+		c.JSON(status, response)
+		return
+	}
+
+	token, refreshToken, err := helper.GenerateAllTokens(createUserInput[0].Email, createUserInput[0].FirstName, createUserInput[0].LastName, createUserInput[0].Role, createUserInput[0].ID)
+	if err != nil {
+		log.Println("ERROR: ", err.Error())
+		status = http.StatusInternalServerError
+		response.Message = "Internal server error."
+		response.Status = status
+		c.JSON(status, response)
+		return
```

Comment:
CodeCov note - this path is not covered since the GenerateAllTokens method doesn't leverage custom logic. It simply uses the `jwt` package to sign and generate some tokens for the user. If an error occurs here, it means something happened within the `jwt` package beyond our control.